DM-6255: Fill out DRP ImChar/JointCal pipeline desciptions in LDM-151

TallJimbo opened the pull request.

TallJimbo committed:
Start fleshing out DRP image characterization.

TallJimbo committed:
Add pseudocode block for BootstrapImChar.

TallJimbo committed:
Update BootstrapImChar pseudocode, data products.

TallJimbo committed:
First complete draft of BootstrapImChar done.

TallJimbo committed:
Write RefineImChar and InitialJointCal sections

BootstrapJointCal and RefineJointCal have been merged into
InitialJointCal.  This still needs to be reflected in diagrams.

TallJimbo committed:
Need to detect, deblend, measure WF sensors.

TallJimbo committed:
Finish JointCal/ImChar section text.

Diagrams still need work, as noted in TODO note.

jdswinbank committed:
Build depends on latex-xcolor.

jdswinbank commented on sections/data_release_production_pipelines.tex:
```diff
@@ -71,26 +75,332 @@ \section{Data Release Production}
+
+The BootstrapImChar pipeline is the first thing run on each science exposure in a data release.  It has the difficult task of bootstrapping multiple quantities (PSF, WCS, photometric calibration, background model, etc.) that each normally require all of the others to be specified when one is fit.  As a result, while the algorithmic components to be run in this pipeline are generally clear, their  ...
+
+A plausible (but by no means certain) high-level algorithm for this pipeline is given below in pseudocode.  Highlighted terms are described in more detail below the pseudocode block.
```
A trivial point, but -- although this is now building on Travis, neither the Travis generated PDF nor one I produced on my laptop have any noticeable highlighting in this block.

jdswinbank commented on sections/data_release_production_pipelines.tex:
```diff
@@ -71,26 +75,332 @@ \section{Data Release Production}
+    sources[ccd] = %\hr{DeblendSources}%(sources[ccd], calexp.{mi,psf})
+    sources[ccd] = %\hr{MeasureSources}%(sources[ccd], calexp.{mi,psf})
+    matches = %\hr{MatchSemiBlind}%(sources, reference)
+    while not converged:
```
Define "converged". Is it guaranteed within $SMALLNUM iterations? Can we set an upper limit on the number of iterations required, and if we hit it either drop the data or press on regardless?

jdswinbank commented on sections/data_release_production_pipelines.tex:
```diff
@@ -71,26 +75,332 @@ \section{Data Release Production}
+\paragraph{RunISR}
+\label{sec:drpBootstrapImChar_RunISR}
+
+Delegate to the \hyperref[sec:acISR]{ISR algorithmic component} to perform standard detrending as well as brighter-fatter correction and interpolation for pixel-area variations (\hyperref[sec:acFixPixelAreaVariations]{Warping Irregularly-Sampled Images}).  It is possible that these corrections will require a PSF model, and hence must be backed-out and recorrected at a later stage when an improved ...
```
ISR requires additional inputs, of course, which we don't call out as inputs to BoostrapImChar. That makes sense: we're not feeding these inputs through the pipeline in the same way as the data, and we don't want to tightly couple this section of the docs to the details of the calibration products/ISR procedure. It would be useful to call it out in some way, though: ultimately, we'll want to be able to quickly verify all the inputs and outputs of everything without digging down to Algorithmic Components.

jdswinbank committed:
Do not run pdfunite.

jdswinbank commented on sections/data_release_production_pipelines.tex:
```diff
@@ -71,26 +75,332 @@ \section{Data Release Production}
+\item We match all sources and the reference catalog by delegating to \hyperref[sec:acJointCalMatching]{JointCalMatching}.  This is a non-blind search; we assume the WCSs output by \hyperref[sec:drpBootstrapImChar]{BootstrapImChar} are good enough that we don't need to fit for any additional offsets between images at this stage.  Some matches will not include a reference object, as the sources wi ...
+\item We classify matches to select a clean sample of low-variability stars for later steps, delegating to \hyperref[sec:acJointCalClassification]{JointCalClassification}.  This uses morphological and possibly color information from source measurements as well as reference catalog information (where available).  This step also assigns an inferred SED to each match from its colors; for matches ass ...
+\item We fit simultaneously for improved astrometric solution by requiring each star in a match to have the same position.  This may need to correct (perhaps approximately) for centroid shifts due to DCR and/or proper motion; if it does not, it must be robust against these shifts (perhaps via outlier rejection).  The models and parameters to fit are TBD in detail, but they will represent further  ...
+\item We fit simultaneously for photometric zeropoints by requiring each star in a match to have the same flux after applying smoothed monochromatic flat fields produced by the calibration products pipeline.  There is a small chance this fit will also be used to further constrain those monochromatic flat fields.  This fit generates a new PhotoCalib component for each CalExp.
```
What we're calling `PhotoCalib` here is (presumably) what we were calling `Calib` earlier.

jdswinbank replied:
Smoothing of flats carried out here or in the CPP? Is this smoothing the same as the iteration described on [page 5 of Robert's easy document](https://github.com/lsst-dm/calibration/blob/master/calibrationPipelines.pdf) which results in "smaller and smoother correction"?

TallJimbo replied:
Yes.  I've tried to switch to `PhotoCalib` to disambiguate, but didn't catch all the places I was using `Calib` (hopefully I have in the latest commit).

The smoothed flats are item 10 of appendix C in that document; I believe the smoothing is done by CPP, but I'm not sure it's the same smoothing described on page 5.

TallJimbo committed:
Clarify "TBDs" in DRP section.

jdswinbank commented on sections/data_release_production_pipelines.tex:
```diff
@@ -71,26 +75,332 @@ \section{Data Release Production}
+    sources[ccd] = %\hr{MeasureSources}%(sources[ccd], calexp.{mi,psf})
+    matches = %\hr{MatchSemiBlind}%(sources, reference)
+    while not converged:
+        %\hr{SelectStars}%(matches, exposures)
```
Is the lack of obvious output from `SelectStars` intended to indicate that it's updating `matches` in place?

jdswinbank commented on sections/data_release_production_pipelines.tex:
```diff
@@ -71,26 +75,332 @@ \section{Data Release Production}
+As we iterate in crowded fields, the number of degrees of freedom in the PSF model should be allowed to slowly increase.
+
+\paragraph{WriteDiagnostics}
+\label{sec:drpBootstrapImChar_WriteDiagnostics}
```
I'm wondering what the motivation in calling this out specifically is. I can see that it falls at a point in the processing where one might well well want diagnostics, but it's not obvious that the same doesn't apply elsewhere.

TallJimbo committed:
Use PhotoCalib, not Calib, for Exposure component

We're currently using "Calib" in the codebase, but this document is
already too full of subtly different meanings of "calibration"; better
to be a bit more explicit.

jdswinbank commented on sections/data_release_production_pipelines.tex:
```diff
@@ -71,26 +75,332 @@ \section{Data Release Production}
+
+\begin{itemize}
+\item The background is not re-estimated and subtracted.
+\item The suite of plugin run by \hyperref[sec:acSingleVisitMeasurement]{Single Visit Measurement} is expanded to included all algorithms indicated in the first column of Figure~\ref{fig:measurement-matrix}.  This should provide all measurements in the DPDD Source table description.
```
`fig:measurement-matrix` is pretty hard to find based on hyperlink if reading on paper. Since it's a long way away in the document, maybe worth adding a page reference.

TallJimbo replied:
I know we decided to go easy on the macros, but I think it'd be quite helpful to have macro that provided both linked text and a section number, while fitting into a naming scheme for sections that would let the label to reference be determined from the link text.  That's a level of standardization I clearly don't have time to do now (and I think it goes with coming up with CamelCase or otherwise code-like names for everything, which no one else seems to be doing though I've heard some enthusiastic noises).
